Feat: article about mobile integration (#29)
Novage/corporate-site: src/content/blog/Integrate-P2P-Video-streaming-into-Mobile-Application.md

---
canonicalURL: https://novage.com.ua/blog/setting-up-P2P-video-in-mobile-app-for-free
author: Dmytro Demchenko
date: "2025-01-29"

> title: "Integrate P2P Video streaming into Mobile Applications: Native Android and Flutter Approaches"

Reviewer: ```suggestion
title: "Integrate P2P Video Streaming into Mobile Applications: Android and iOS Approaches"
```

description: "Discover how to integrate P2P Media Loader into your mobile applications to achieve scalable, cost-efficient video streaming. This article explores two practical approaches — Native Android and Flutter — to leverage peer-to-peer (P2P) streaming technology powered by WebRTC. Learn step-by-step integration techniques, review the pros and cons of each method."
---

In today’s rapidly evolving digital landscape, delivering high-quality video content seamlessly and cost-effectively is essential for modern mobile applications. Traditional Content Delivery Networks (CDNs) often face scalability challenges and incur significant bandwidth costs, especially as user bases grow. This is where Peer-to-Peer (P2P) streaming technology steps in.

> P2P streaming leverages WebRTC to enable devices to share video data directly with one another, thereby reducing the load on centralized servers. This approach not only cuts down bandwidth expenses but also enhances streaming performance and scalability. Whether you're developing an on-demand video service, or live-streaming platform, P2P streaming offers a robust alternative that can lead to smoother playback experiences and lower operational costs.

Reviewer: ```suggestion
P2P streaming enables devices to share video data directly with one another, thereby reducing the load on centralized servers. This approach not only cuts down bandwidth expenses but also enhances streaming performance and scalability. Whether you're developing an on-demand video service, or live-streaming platform, P2P streaming offers a robust alternative that can lead to smoother playback experiences and lower operational costs.
```

> Enter Peer-to-Peer (P2P) Streaming Technology with [P2P Media Loader](https://github.com/Novage/p2p-media-loader) — an open-source JavaScript library that leverages modern web browser features to enable media delivery over peer-to-peer (P2P) connections. The library enables the creation of a huge P2P mesh networks, also known as peer-to-peer content delivery network (P2P CDN), peer-to-peer television (P2PTV), and Enterprise Content Delivery Network (eCDN), which allows traffic sharing among users who are simultaneously viewing the same live or video on demand (VOD) stream via HLS or MPEG-DASH protocols.

Reviewer: ```suggestion
Enter Peer-to-Peer Streaming Technology with [P2P Media Loader](https://github.com/Novage/p2p-media-loader) — an open-source JavaScript library that leverages modern web browser features to enable media delivery over peer-to-peer connections. The library enables the creation of a huge P2P mesh networks, also known as peer-to-peer content delivery network (P2P CDN), peer-to-peer television (P2PTV), and Enterprise Content Delivery Network (eCDN), which allows traffic sharing among users who are simultaneously viewing the same live or video on demand (VOD) stream via HLS or MPEG-DASH protocols.
```

> In this article, we will show you how to integrate P2P Media Loader into your mobile application. Integrations are fully compatible with web-based solutions, meaning that if you have both web and mobile integrations, your users can share traffic across platforms when watching the same content. As a result, you'll build a larger network of peers, save more on traffic costs, and enjoy enhanced streaming performance.

Reviewer: We will also show both iOS and Android integration samples that are available for now.

> In this article, we will show you how to integrate P2P Media Loader into your mobile application. Integrations are fully compatible with web-based solutions, meaning that if you have both web and mobile integrations, your users can share traffic across platforms when watching the same content. As a result, you'll build a larger network of peers, save more on traffic costs, and enjoy enhanced streaming performance.

Reviewer: Say that with these approaches mobile (iOS and Android) peers are fully compatible with web peers organizing all of them in one P2P network.

- **Better Control Over Media Playback:** Native players like ExoPlayer offer better control over media streaming.
- **Access to Latest Android Features:** Native development ensures immediate access to the latest Android platform updates and features without waiting for framework support.

**Cons:**

- **Platform-Specific:** The native approach is limited to Android.
- **Dependency on WebView:** The library internally relies on WebView, it may introduce performance overhead.

In this section, we introduce our [Kotlin library](https://github.com/Novage/p2p-media-loader-mobile) designed to leverage the capabilities of [P2P Media Loader](https://github.com/novage/p2p-media-loader) and WebRTC for efficient peer-to-peer streaming.

Our native Android library is crafted in Kotlin and is fully compatible with [ExoPlayer](https://exoplayer.dev/), one of the most popular media players for Android. The library is also designed to support other media players with minimal adjustments.

**How the library works?**

![Kotlin Library Architecture](../../assets/kotlin-diagram.png)

> Inside the library we launch embedded server that will respond on players request. To setup P2P Media Loader we decided to use a WebView without binding to a view. The approach with WebView allows us to maintain the same codebase across different platforms. Moreover, with our approach there is no need to write custom WebRTC implementation for Native Android, since in WebView we already have WebRTC. This means that our WEB integration of P2P Media Loader is fully compatible with Native Android integration, as a result consumers may have a bigger mesh of peers to efficiently transfer traffic between each other.

Reviewer: ```suggestion
Inside the library we launch embedded HTTP server that will respond on players request. To setup P2P Media Loader we decided to use an off-screen WebView. This approach is lightwait because WebView is a part of OS and no need to bundle external libraries. Also this means that our WEB integration of P2P Media Loader is fully compatible with Native Android integration, as a result consumers may have a bigger mesh of peers to efficiently transfer traffic between each other.
```

            .createMediaSource(mediaItem)

        exoPlayer.apply {
            playWhenReady = true
            setMediaSource(mediaSource)
            prepare()
        }
    }

    private fun onP2PReadyErrorCallback(error: String) {
        // Implement error handling logic here
    }

    override fun onStop() {
        super.onStop()
        // Disable P2P features when the activity stops
        // See https://novage.github.io/p2p-media-loader/docs/v2.1.0/types/p2p-media-loader-core.DynamicCoreConfig.html
        p2pml.applyDynamicConfig("{\"isP2PDisabled\": true}")
    }

    override fun onRestart() {
        super.onRestart()
        // Re-enable P2P features when the activity restarts
        // See https://novage.github.io/p2p-media-loader/docs/v2.1.0/types/p2p-media-loader-core.DynamicCoreConfig.html
        p2pml.applyDynamicConfig("{\"isP2PDisabled\": false}")
    }

    override fun onDestroy() {
        super.onDestroy()
        // Release ExoPlayer resources and stop P2P Media Loader
        exoPlayer.release()
        p2pml.stop()
    }
}
```

More examples are available in the library [repository](https://github.com/Novage/p2p-media-loader-mobile)

> ## Flutter Integration

Reviewer: Should go first
```suggestion
## Flutter Integration (iOS and Android)
```

                  // Enable to debug WebView
                  isInspectable: true,
                ),
                onWebViewCreated: _onWebViewCreated,
              ))),
    );
  }
}
```

**For a more extensive example that gathers P2P engine stats and manages P2P state according to app lifecycle, see our [P2P Media Loader Flutter Demo](https://github.com/Novage/p2p-media-loader-flutter-demo)**

## Testing P2P with Other Peers

**Important Note:** P2P (WebRTC) may not connect to the outer world if it runs on Android emulators due to its virtual machine [network configuration](https://developer.android.com/studio/run/emulator-networking) (NAT). Please test P2P connectivity on real devices.

After integrating the P2P Media Loader into your Flutter application, it's crucial to verify that the peer-to-peer connections are functioning as intended.

The simpliest way to test is to run the application with integrated P2P on a real device and open our [demo](https://novage.com.ua/p2p-media-loader/demo) with same manifest url.

**For Production Use:**

- Setting up a production WebTorrent tracker.
- Customizing RTC configurations to suit your network conditions.

> - Specifying other configuration parameters such as **swarmId** and other settings required for optimal performance,

Reviewer: ```suggestion
- Specifying other configuration parameters such as **swarmId** and other settings required for optimal performance.
```

> For detailed guidance on production configurations and advanced customizations options, please refer to our [documentation](https://novage.github.io/p2p-media-loader/docs/latest/).

Reviewer: ```suggestion
For detailed guidance on production configurations and advanced customizations options, please refer to our GitHub and [documentation](https://novage.github.io/p2p-media-loader/docs/latest/).
```
